torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1332,6 +1349,9 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # the transfer is async so it is not necessarily done until we
                 # explicitly synchronize in backward.
                 param.grad.data = param._cpu_grad  # type: ignore[attr-defined]
+                # We ensure that `param._saved_grad_shard` is on the same device
+                # as `param.grad` (to appease autograd).
+                param._saved_grad_shard = param._cpu_grad  # type: ignore[attr-defined]
```

Comment:
based on above comments, we should move this to final callback of main backward pass